File: algorithms/code/leetcode/lc154_find_minimum_in_rotated_sorted_array_ii/lc154_find_minimum_in_rotated_sorted_array_ii.py
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/

import unittest


class Solution:
    def findMin(self, nums: list[int]) -> int:
        a = 0
        z = len(nums) - 1
        while a < z:
            m = (a + z) >> 1
            if nums[z] < nums[m]:
                a = m + 1
            elif nums[m] < nums[z]:
                z = m
            else:
                z -= 1
        return nums[a]


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        nums = [4, 5, 6, 7, 0, 1, 2]
        assert Solution().findMin(nums) == 0

    def test_trickier(self) -> None:
        nums = [1, 1, 0, 1, 1, 1, 1]
        assert Solution().findMin(nums) == 0

File: algorithms/code/leetcode/lc154_find_minimum_in_rotated_sorted_array_ii/lc154_find_minimum_in_rotated_sorted_array_ii.rs
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/
// #hard

pub struct Solution;

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let mut a = 0;
        let mut z = nums.len() - 1;
        while a < z {
            let m = (a + z) / 2;
            match nums[z].cmp(&nums[m]) {
                std::cmp::Ordering::Less => {
                    a = m + 1;
                }
                std::cmp::Ordering::Greater => {
                    z = m;
                }
                _ => {
                    z -= 1;
                }
            }
        }
        nums[a]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let nums = vec![4, 5, 6, 7, 0, 1, 2];
        assert_eq!(Solution::find_min(nums), 0);
    }

    #[test]
    fn test_trickier() {
        let nums = vec![1, 1, 0, 1, 1, 1, 1];
        assert_eq!(Solution::find_min(nums), 0);
    }
}
